include/kernel.h

```diff
@@ -0,0 +1,531 @@
+ * for constants and macros that simplify conversion to standard time units.
+ *
+ * Notice that #wait_ticks and #wait_ms are not threading aware, so they will
+ * just spin-wait while keeping the current thread ready. Other higher-priority
+ * threads might still be scheduled if they are awaken by an interrupt, but in
+ * general thread_sleep is a better primitive if you want the current thread
+ * to yield.
+ *
+ * @note This function requires the timer module, so #timer_init must have
+ *       been called.
+ *
+ * @param[in] ticks
+ *            The number of hardware ticks to sleep.
+ */
+void kthread_sleep(uint32_t ticks);
+
+/** 
+ * @brief Change priority of a thread.
+ * 
+ * Change priority of the specified thread. If the argument is NULL,
+ * this function changes priority to the current thread.
+ *
+ * The change of priority is immediately effective. It may cause
+ * a context switch if the changed thread is ready and its priority
+ * is changed in a way to start/stop it relative to the other ready
+ * threads.
+ *
+ * @param[in]  th
+ *             Reference to the thread (NULL = current thread)
+ * @param[in]  pri
+ *             New priority of the thread.
+ *
+ */
+void kthread_set_pri(kthread_t *th, int pri);
+
+
+/** 
+ * @brief Kill a thread, aborting its execution.
+ *
+ * The specified thread is aborted, and its memory freed (including
+ * its stack). This function can executed for any thread, including the
+ * current one. The execution will be aborted and the memory released
+ * to the heap for further use.
+ *
+ * @param[in]  th
+ *             Reference to the thread to kill (NULL = current thread)
+ *
+ */
+void kthread_kill(kthread_t *th);
+
+
+/**
+ * @brief Return the name of the specified thread.
+ *
+ * @param[in]  th
+ *             Reference to thread (NULL = current thread)
+ */
+const char* kthread_name(kthread_t *th);
+
+/**
+ * @brief Wait for a single hardware event.
+ *
+ * This functions blocks the current thread until a specified
+ * hardware event happens. The CPU is yielded so that other
+ * threads will be scheduled.
+ *
+ * If the thread needs to handle multiple events and/or a more
+ * elaborate communication with other threads, you can use
+ * #thread_listen_event instead.
+ *
+ * @param[in] evt
+ *            The event to wait for
+ *
+ */
+void kthread_wait_event(kevent_t *evt);
+
+/**
+ * @brief Allocate a new mailbox (on the heap).
+ *
+ * @param[in] max_mails
+ *            Maximum number of pending mails in the mailbox.
+ *
+ * @return The allocated mailbox
+ *
+ * @note See #mbox_new_stack to allocate a mailbox on the stack.
+ */
+kmbox_t* kmbox_new(int max_mails);
+
+/**
+ * @brief Allocate a new mailbox (on the stack).
+ *
+ * @param[in] max_mails
+ *            Maximum number of pending mails in the mailbox.
+ *
+ * @return The allocated mailbox
+ *
+ * @note See #mbox_new to allocate a mailbox on the heap.
+ */
+#define kmbox_new_stack(max_mails) ({ \
```

Comment:
At a later point, it might be good to have some sample code for `kmbox_new_stack` for new programmers that aren't as familiar with the return value of a C macro. They trip me up sometimes!

Reply:
There's an example in the documentation of `kmbox_attach_event`.